File: src/source/ast.scala
/**
  * Copyright 2014 Dropbox, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *    http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  * 
  * This file has been modified by Snap, Inc.
  */

package djinni.ast

import java.io.File
import djinni.ast.Record.DerivingType.DerivingType
import djinni.meta.MExpr
import djinni.syntax.Loc

case class IdlFile(imports: Seq[FileRef], typeDecls: Seq[TypeDecl], flags: Seq[String])

abstract sealed class FileRef {
  val file: File
}
case class IdlFileRef(override val file: File) extends FileRef
case class ExternFileRef(override val file: File) extends FileRef
case class ProtobufFileRef(override val file: File) extends FileRef

case class Ident(name: String, file: File, loc: Loc)
class ConstRef(ident: Ident) extends Ident(ident.name, ident.file, ident.loc)
class EnumValue(val ty: Ident, ident: Ident) extends Ident(ident.name, ident.file, ident.loc)

case class TypeParam(ident: Ident)

case class Doc(lines: Seq[String])

sealed abstract class TypeDecl {
  val ident: Ident
  val params: Seq[TypeParam]
  val body: TypeDef
  val origin: String
}
case class InternTypeDecl(override val ident: Ident, override val params: Seq[TypeParam], override val body: TypeDef, doc: Doc, override val origin: String) extends TypeDecl
case class ExternTypeDecl(override val ident: Ident, override val params: Seq[TypeParam], override val body: TypeDef, properties: Map[String, Any], override val origin: String) extends TypeDecl
case class ProtobufTypeDecl(override val ident: Ident, override val params: Seq[TypeParam], override val body: TypeDef, override val origin: String) extends TypeDecl

case class Ext(java: Boolean, cpp: Boolean, objc: Boolean, js: Boolean) {
  def any(): Boolean = {
    java || cpp || objc || js
  }
}

case class TypeRef(expr: TypeExpr) {
  var resolved: MExpr = null
}
case class TypeExpr(ident: Ident, args: Seq[TypeExpr])

sealed abstract class TypeDef

case class Const(ident: Ident, ty: TypeRef, value: Any, doc: Doc)

case class Enum(options: Seq[Enum.Option], flags: Boolean) extends TypeDef
object Enum {
  object SpecialFlag extends Enumeration {
    type SpecialFlag = Value
    val NoFlags, AllFlags = Value
  }
  import SpecialFlag._
  case class Option(ident: Ident, doc: Doc, specialFlag: scala.Option[SpecialFlag])
}

case class Record(ext: Ext, fields: Seq[Field], consts: Seq[Const], derivingTypes: Set[DerivingType], baseRecord: scala.Option[String]) extends TypeDef
object Record {
  object DerivingType extends Enumeration {
    type DerivingType = Value
    val Ord, AndroidParcelable, NSCopying = Value
  }
}

case class Interface(ext: Ext, methods: Seq[Interface.Method], consts: Seq[Const]) extends TypeDef
object Interface {
  case class Method(ident: Ident, params: Seq[Field], ret: Option[TypeRef], doc: Doc, static: Boolean, const: Boolean, lang: Ext)
}

case class Field(ident: Ident, ty: TypeRef, defaultValue: String, doc: Doc)

case class SuperRecord(ident: Ident, record: Record, fields: Seq[Field])

case class ProtobufMessage(cpp: ProtobufMessage.Cpp, java: ProtobufMessage.Java, objc: Option[ProtobufMessage.Objc], ts: Option[ProtobufMessage.Ts]) extends TypeDef
object ProtobufMessage {
  case class Cpp(header: String, ns: String)
  case class Java(pkg: String, jniClass: Option[String], jniHeader: Option[String])
  case class Objc(header: String, prefix: String)
  case class Ts(module: String, ns: String)
}

File: djinni-rs/crates/djinni-ast/src/ast.rs
// vinsg
// AST types for the Djinni IDL, translated from ast.scala

use std::collections::BTreeSet;
use std::path::PathBuf;

use crate::meta::MExpr;

#[derive(Debug, Clone)]
pub struct Loc {
    pub file: PathBuf,
    pub line: usize,
    pub col: usize,
}

impl std::fmt::Display for Loc {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} ({}.{})", self.file.display(), self.line, self.col)
    }
}

#[derive(Debug, Clone)]
pub struct Ident {
    pub name: String,
    pub loc: Loc,
}

#[derive(Debug, Clone)]
pub struct TypeParam {
    pub ident: Ident,
}

#[derive(Debug, Clone)]
pub struct Doc {
    pub lines: Vec<String>,
}

// File references
#[derive(Debug, Clone)]
pub enum FileRef {
    Idl(PathBuf),
    Extern(PathBuf),
    Protobuf(PathBuf),
}

impl FileRef {
    pub fn file(&self) -> &PathBuf {
        match self {
            FileRef::Idl(f) | FileRef::Extern(f) | FileRef::Protobuf(f) => f,
        }
    }
}

#[derive(Debug, Clone)]
pub struct IdlFile {
    pub imports: Vec<FileRef>,
    pub type_decls: Vec<TypeDecl>,
    pub flags: Vec<String>,
}

// Type declarations
#[derive(Debug, Clone)]
pub enum TypeDecl {
    Intern {
        ident: Ident,
        params: Vec<TypeParam>,
        body: TypeDef,
        doc: Doc,
        origin: String,
    },
    Extern {
        ident: Ident,
        params: Vec<TypeParam>,
        body: TypeDef,
        properties: std::collections::HashMap<String, serde_yaml::Value>,
        origin: String,
    },
    Protobuf {
        ident: Ident,
        params: Vec<TypeParam>,
        body: TypeDef,
        origin: String,
    },
}

impl TypeDecl {
    pub fn ident(&self) -> &Ident {
        match self {
            TypeDecl::Intern { ident, .. }
            | TypeDecl::Extern { ident, .. }
            | TypeDecl::Protobuf { ident, .. } => ident,
        }
    }

    pub fn params(&self) -> &[TypeParam] {
        match self {
            TypeDecl::Intern { params, .. }
            | TypeDecl::Extern { params, .. }
            | TypeDecl::Protobuf { params, .. } => params,
        }
    }

    pub fn body(&self) -> &TypeDef {
        match self {
            TypeDecl::Intern { body, .. }
            | TypeDecl::Extern { body, .. }
            | TypeDecl::Protobuf { body, .. } => body,
        }
    }

    pub fn origin(&self) -> &str {
        match self {
            TypeDecl::Intern { origin, .. }
            | TypeDecl::Extern { origin, .. }
            | TypeDecl::Protobuf { origin, .. } => origin,
        }
    }

    pub fn body_mut(&mut self) -> &mut TypeDef {
        match self {
            TypeDecl::Intern { body, .. }
            | TypeDecl::Extern { body, .. }
            | TypeDecl::Protobuf { body, .. } => body,
        }
    }
}

// Language extension flags
#[derive(Debug, Clone, Default)]
pub struct Ext {
    pub java: bool,
    pub cpp: bool,
    pub objc: bool,
    pub js: bool,
}

impl Ext {
    pub fn any(&self) -> bool {
        self.java || self.cpp || self.objc || self.js
    }
}

// Type references (resolved during type resolution)
#[derive(Debug, Clone)]
pub struct TypeRef {
    pub expr: TypeExpr,
    pub resolved: Option<MExpr>,
}

#[derive(Debug, Clone)]
pub struct TypeExpr {
    pub ident: Ident,
    pub args: Vec<TypeExpr>,
}

// Type definitions
#[derive(Debug, Clone)]
pub enum TypeDef {
    Enum(Enum),
    Record(Record),
    Interface(Interface),
    ProtobufMessage(ProtobufMessage),
}

// Constants
#[derive(Debug, Clone)]
pub struct Const {
    pub ident: Ident,
    pub ty: TypeRef,
    pub value: ConstValue,
    pub doc: Doc,
}

#[derive(Debug, Clone)]
pub enum ConstValue {
    Int(i64),
    Float(f64),
    Bool(bool),
    String(String),
    EnumValue { ty: String, value: String },
    Composite(Vec<(String, ConstValue)>),
}

// Enum definition
#[derive(Debug, Clone)]
pub struct Enum {
    pub options: Vec<EnumOption>,
    pub flags: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SpecialFlag {
    NoFlags,
    AllFlags,
}

#[derive(Debug, Clone)]
pub struct EnumOption {
    pub ident: Ident,
    pub doc: Doc,
    pub special_flag: Option<SpecialFlag>,
}

// Record definition
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum DerivingType {
    Ord,
    AndroidParcelable,
    NSCopying,
}

#[derive(Debug, Clone)]
pub struct Record {
    pub ext: Ext,
    pub fields: Vec<Field>,
    pub consts: Vec<Const>,
    pub deriving_types: BTreeSet<DerivingType>,
    pub base_record: Option<String>,
}

// Interface definition
#[derive(Debug, Clone)]
pub struct Interface {
    pub ext: Ext,
    pub methods: Vec<Method>,
    pub consts: Vec<Const>,
}

#[derive(Debug, Clone)]
pub struct Method {
    pub ident: Ident,
    pub params: Vec<Field>,
    pub ret: Option<TypeRef>,
    pub doc: Doc,
    pub is_static: bool,
    pub is_const: bool,
    pub lang: Ext,
}

// Field definition
#[derive(Debug, Clone)]
pub struct Field {
    pub ident: Ident,
    pub ty: TypeRef,
    pub default_value: String,
    pub doc: Doc,
}

// Super record (for inheritance)
#[derive(Debug, Clone)]
pub struct SuperRecord {
    pub ident: Ident,
    pub record: Record,
    pub fields: Vec<Field>,
}

// Protobuf message
#[derive(Debug, Clone)]
pub struct ProtobufMessage {
    pub cpp: ProtobufCpp,
    pub java: ProtobufJava,
    pub objc: Option<ProtobufObjc>,
    pub ts: Option<ProtobufTs>,
}

#[derive(Debug, Clone)]
pub struct ProtobufCpp {
    pub header: String,
    pub ns: String,
}

#[derive(Debug, Clone)]
pub struct ProtobufJava {
    pub pkg: String,
    pub jni_class: Option<String>,
    pub jni_header: Option<String>,
}

#[derive(Debug, Clone)]
pub struct ProtobufObjc {
    pub header: String,
    pub prefix: String,
}

#[derive(Debug, Clone)]
pub struct ProtobufTs {
    pub module: String,
    pub ns: String,
}
